src/core/ddsi/src/ddsi_cdrstream.c

```diff
@@ -521,182 +608,464 @@ static const uint32_t *find_union_case (const uint32_t * __restrict union_ops, u
       return ops + (jmp ? jmp : 5);
     }
+    case DDS_OP_VAL_EXT: {
+      abort (); /* not supported */
+      break;
+    }
   }
   return NULL;
 }
 
-static const uint32_t *dds_stream_write_uni (dds_ostream_t * __restrict os, const char * __restrict discaddr, const char * __restrict baseaddr, const uint32_t * __restrict ops, uint32_t insn)
+static const uint32_t *skip_union_default (uint32_t insn, char * __restrict discaddr, char * __restrict baseaddr, const uint32_t * __restrict ops)
 {
-  const uint32_t disc = write_union_discriminant (os, DDS_OP_SUBTYPE (insn), discaddr);
-  uint32_t const * const jeq_op = find_union_case (ops, disc);
+  switch (DDS_OP_SUBTYPE (insn))
+  {
+    case DDS_OP_VAL_1BY: *((uint8_t *) discaddr) = 0; break;
+    case DDS_OP_VAL_2BY: *((uint16_t *) discaddr) = 0; break;
+    case DDS_OP_VAL_4BY: case DDS_OP_VAL_ENU: *((uint32_t *) discaddr) = 0; break;
+    default: break;
+  }
+  uint32_t const * const jeq_op = find_union_case (ops, 0);
   ops += DDS_OP_ADR_JMP (ops[3]);
   if (jeq_op)
   {
     const enum dds_stream_typecode valtype = DDS_JEQ_TYPE (jeq_op[0]);
-    const void *valaddr = baseaddr + jeq_op[2];
+    void *valaddr = baseaddr + jeq_op[2];
     switch (valtype)
     {
-      case DDS_OP_VAL_1BY: dds_os_put1 (os, *(const uint8_t *) valaddr); break;
-      case DDS_OP_VAL_2BY: dds_os_put2 (os, *(const uint16_t *) valaddr); break;
-      case DDS_OP_VAL_4BY: dds_os_put4 (os, *(const uint32_t *) valaddr); break;
-      case DDS_OP_VAL_8BY: dds_os_put8 (os, *(const uint64_t *) valaddr); break;
-      case DDS_OP_VAL_STR: dds_stream_write_string (os, *(const char **) valaddr); break;
-      case DDS_OP_VAL_BST: dds_stream_write_string (os, (const char *) valaddr); break;
-      case DDS_OP_VAL_SEQ: case DDS_OP_VAL_ARR: case DDS_OP_VAL_UNI: case DDS_OP_VAL_STU:
-        dds_stream_write (os, valaddr, jeq_op + DDS_OP_ADR_JSR (jeq_op[0]));
+      case DDS_OP_VAL_1BY: *((uint8_t *) valaddr) = 0; break;
+      case DDS_OP_VAL_2BY: *((uint16_t *) valaddr) = 0; break;
+      case DDS_OP_VAL_4BY: case DDS_OP_VAL_ENU: *((uint32_t *) valaddr) = 0; break;
+      case DDS_OP_VAL_8BY: *((uint64_t *) valaddr) = 0; break;
+      case DDS_OP_VAL_STR: *(char **) valaddr = dds_stream_reuse_string_empty (*((char **) valaddr)); break;
+      case DDS_OP_VAL_BST: case DDS_OP_VAL_BSP: case DDS_OP_VAL_SEQ: case DDS_OP_VAL_ARR: case DDS_OP_VAL_UNI: case DDS_OP_VAL_STU:
+        (void) dds_stream_skip_default (valaddr, jeq_op + DDS_OP_ADR_JSR (jeq_op[0]));
         break;
+      case DDS_OP_VAL_EXT: {
+        abort (); /* not supported */
+        break;
+      }
     }
   }
   return ops;
 }
 
-static void dds_stream_write (dds_ostream_t * __restrict os, const char * __restrict data, const uint32_t * __restrict ops)
+static const uint32_t *dds_stream_write_delimited (dds_ostream_t * __restrict os, const char * __restrict data, const uint32_t * __restrict ops)
+{
+  dds_os_put4 (os, 0xffffffff);
```

Comment:
Instead of actually writing a 32-bit value to the stream, would it not make more sense to add a function to skip 4 (aligned) bytes and return the offset of those 4 bytes? I.e., something like `const uint32_t offs = dds_os_reserve4 (os);`?